File: src/main/kotlin/com/github/shiraji/findpullrequest/menu/FindPullRequestMenu.java
package com.github.shiraji.findpullrequest.menu;

import com.github.shiraji.findpullrequest.model.FindPullRequestConfig;
import com.github.shiraji.findpullrequest.model.FindPullRequestHostingServices;
import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class FindPullRequestMenu implements Configurable {
    private JCheckBox disable;
    private JCheckBox debugMode;
    private JCheckBox jumpToFile;
    private JComboBox<Protocol> protocols;
    private JPanel root;
    private JCheckBox copyPopup;
    private JComboBox<FindPullRequestHostingServices> hostingService;
    private PropertiesComponent config;

    private enum Protocol {
        https("https://"), http("http://");
        private String text;

        Protocol(String text) {
            this.text = text;
        }

        @NotNull
        public String getText() {
            return text;
        }

        @Nullable
        public static Protocol findProtocolBy(String name) {
            switch (name) {
                case "https://":
                    return https;
                case "http://":
                    return http;
                default:
                    return null;
            }
        }
    }

    FindPullRequestMenu(Project project) {
        super();

        config = PropertiesComponent.getInstance(project);
        protocols.setModel(new DefaultComboBoxModel<>(Protocol.values()));
        hostingService.setModel(new DefaultComboBoxModel<>(FindPullRequestHostingServices.values()));
    }

    @Nls
    @Override
    public String getDisplayName() {
        return "Find Pull Request Plugin";
    }

    @Nullable
    @Override
    public String getHelpTopic() {
        return null;
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        return root;
    }

    @Override
    public boolean isModified() {
        return isModifiedProtocol()
                || isModifiedHostingService()
                || FindPullRequestConfig.isDebugMode(config) != debugMode.isSelected()
                || FindPullRequestConfig.isJumpToFile(config) != jumpToFile.isSelected()
                || FindPullRequestConfig.isDisable(config) != disable.isSelected()
                || FindPullRequestConfig.isPopupAfterCopy(config) != copyPopup.isSelected();
    }

    private boolean isModifiedProtocol() {
        Object selectedItem = protocols.getSelectedItem();
        return selectedItem instanceof Protocol && !FindPullRequestConfig.getProtocol(config).equals(
                ((Protocol)selectedItem).getText());
    }

    private boolean isModifiedHostingService() {
        Object selectedItem = hostingService.getSelectedItem();
        return selectedItem instanceof FindPullRequestHostingServices
                && !FindPullRequestConfig.getHosting(config).equals(((FindPullRequestHostingServices)selectedItem).name());
    }

    @Override
    public void apply() throws ConfigurationException {
        FindPullRequestConfig.setDisable(config, disable.isSelected());
        FindPullRequestConfig.setDebugMode(config, debugMode.isSelected());
        FindPullRequestConfig.setJumpToFile(config, jumpToFile.isSelected());
        FindPullRequestConfig.setPopupAfterCopy(config, copyPopup.isSelected());
        applyProtocol();
        applyHostingService();
    }

    private void applyProtocol() {
        Object selectedItem = protocols.getSelectedItem();
        if (selectedItem instanceof Protocol) {
            FindPullRequestConfig.setProtocol(config, ((Protocol)selectedItem).getText());
        }
    }

    private void applyHostingService() {
        Object selectedItem = hostingService.getSelectedItem();
        if (selectedItem instanceof FindPullRequestHostingServices) {
            FindPullRequestConfig.setHosting(config, (FindPullRequestHostingServices) selectedItem);
        }
    }

    @Override
    public void reset() {
        resetProtocols();
        resetHostingServices();
        disable.setSelected(FindPullRequestConfig.isDisable(config));
        debugMode.setSelected(FindPullRequestConfig.isDebugMode(config));
        jumpToFile.setSelected(FindPullRequestConfig.isJumpToFile(config));
        copyPopup.setSelected(FindPullRequestConfig.isPopupAfterCopy(config));
    }

    private void resetProtocols() {
        String protocol = FindPullRequestConfig.getProtocol(config);
        Protocol protocolBy = Protocol.findProtocolBy(protocol);
        int selectedIndex = 0;
        if (protocolBy != null) {
            selectedIndex = protocolBy.ordinal();
        }
        protocols.setSelectedIndex(selectedIndex);
    }

    private void resetHostingServices() {
        String hostingServiceConf = FindPullRequestConfig.getHosting(config);
        FindPullRequestHostingServices hostingServices = FindPullRequestHostingServices.findBy(hostingServiceConf);
        int selectedIndex = hostingServices.ordinal();
        hostingService.setSelectedIndex(selectedIndex);
    }

    @Override
    public void disposeUIResources() {

    }
}

File: src/main/kotlin/com/github/shiraji/findpullrequest/menu/FindPullRequestMenu.kt
package com.github.shiraji.findpullrequest.menu

import com.github.shiraji.findpullrequest.config.getHosting
import com.github.shiraji.findpullrequest.config.getProtocol
import com.github.shiraji.findpullrequest.config.isDisable
import com.github.shiraji.findpullrequest.config.isJumpToFile
import com.github.shiraji.findpullrequest.config.isPopupAfterCopy
import com.github.shiraji.findpullrequest.config.setDisable
import com.github.shiraji.findpullrequest.config.setHosting
import com.github.shiraji.findpullrequest.config.setJumpToFile
import com.github.shiraji.findpullrequest.config.setPopupAfterCopy
import com.github.shiraji.findpullrequest.config.setProtocol
import com.github.shiraji.findpullrequest.domain.HostingService
import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.project.Project
import org.jetbrains.annotations.Nls
import javax.swing.DefaultComboBoxModel
import javax.swing.JCheckBox
import javax.swing.JComboBox
import javax.swing.JComponent
import javax.swing.JPanel

class FindPullRequestMenu internal constructor(project: Project) : Configurable {
    private var disable: JCheckBox? = null
    private var jumpToFile: JCheckBox? = null
    private var protocols: JComboBox<Protocol>? = null
    private var root: JPanel? = null
    private var copyPopup: JCheckBox? = null
    private var hostingService: JComboBox<HostingService>? = null
    private var config: PropertiesComponent = PropertiesComponent.getInstance(project)

    private val isModifiedProtocol: Boolean
        get() {
            val selectedItem = protocols!!.selectedItem
            return selectedItem is Protocol && config.getProtocol() != selectedItem.text
        }

    private val isModifiedHostingService: Boolean
        get() {
            val selectedItem = hostingService!!.selectedItem
            return selectedItem is HostingService && config.getHosting() != selectedItem.name
        }

    private enum class Protocol constructor(val text: String) {
        https("https://"), http("http://");

        companion object {

            fun findProtocolBy(name: String): Protocol? {
                return when (name) {
                    "https://" -> https
                    "http://" -> http
                    else -> null
                }
            }
        }
    }

    init {
        protocols!!.model = DefaultComboBoxModel(Protocol.values())
        hostingService!!.model = DefaultComboBoxModel(HostingService.values())
    }

    @Nls
    override fun getDisplayName(): String {
        return "Find Pull Request Plugin"
    }

    override fun getHelpTopic(): String? {
        return null
    }

    override fun createComponent(): JComponent? {
        return root
    }

    override fun isModified(): Boolean {
        return (isModifiedProtocol ||
            isModifiedHostingService ||
            config.isJumpToFile() != jumpToFile!!.isSelected ||
            config.isDisable() != disable!!.isSelected ||
            config.isPopupAfterCopy() != copyPopup!!.isSelected)
    }

    @Throws(ConfigurationException::class)
    override fun apply() {
        config.setDisable(disable!!.isSelected)
        config.setJumpToFile(jumpToFile!!.isSelected)
        config.setPopupAfterCopy(copyPopup!!.isSelected)
        applyProtocol()
        applyHostingService()
    }

    private fun applyProtocol() {
        val selectedItem = protocols!!.selectedItem
        if (selectedItem is Protocol) {
            config.setProtocol(selectedItem.text)
        }
    }

    private fun applyHostingService() {
        val selectedItem = hostingService!!.selectedItem
        if (selectedItem is HostingService) {
            config.setHosting(selectedItem)
        }
    }

    override fun reset() {
        resetProtocols()
        resetHostingServices()
        disable!!.isSelected = config.isDisable()
        jumpToFile!!.isSelected = config.isJumpToFile()
        copyPopup!!.isSelected = config.isPopupAfterCopy()
    }

    private fun resetProtocols() {
        val protocol = config.getProtocol()
        val protocolBy = Protocol.findProtocolBy(protocol)
        var selectedIndex = 0
        if (protocolBy != null) {
            selectedIndex = protocolBy.ordinal
        }
        protocols!!.selectedIndex = selectedIndex
    }

    private fun resetHostingServices() {
        val hostingServiceConf = config.getHosting()
        val hostingServices = HostingService.findBy(hostingServiceConf)
        val selectedIndex = hostingServices.ordinal
        hostingService!!.selectedIndex = selectedIndex
    }

    override fun disposeUIResources() {
    }
}
